File: leetcode/Path_Sum.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }
        Queue<TreeNode> nq = new LinkedList<TreeNode>();
        Queue<Integer> sq = new LinkedList<Integer>();
        nq.add(root);
        sq.add(root.val);
        while (!nq.isEmpty()) {
            TreeNode cn = nq.poll();
            int cs = sq.poll();
            if (cn.left == null && cn.right == null) {
                if (cs == sum) return true;
            }
            if (cn.left != null) {
                nq.add(cn.left);
                sq.add(cs + cn.left.val);
            }
            if (cn.right != null) {
                nq.add(cn.right);
                sq.add(cs + cn.right.val);
            }
        }
        return false;
    }
}

File: leetcode/Path_Sum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int target_sum;
    bool findPath(TreeNode *node, int current_sum) {
        if (node == NULL)
            return false;
        current_sum += node->val;
        if (node->left == NULL && node->right == NULL &&  // It must be a leaf
            current_sum == target_sum)
            return true;
        //if (current_sum > target_sum)  // Wrong pruning. Values might be negative.
        //    return false;
        if (findPath(node->left, current_sum))
            return true;
        return findPath(node->right, current_sum);
    }
public:
    bool hasPathSum(TreeNode *root, int sum) {
        target_sum = sum;
        return findPath(root, 0);
    }
};
